sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -0,0 +1,253 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeMap, AttributeSet, NamedExpression, PlanExpression}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+
+object DeduplicateRelations extends Rule[LogicalPlan] {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
```

Comment:
```
override def apply(plan: LogicalPlan): LogicalPlan = if (hasConflictingAttrs(plan)) {
  renewDuplicatedRelations(Nil, plan)._1.resolveOperatorsUp {
    ...
} else {
  plan
}
```
?

Reply:
sounds good

Reply:
sounds good

Reply:
does `hasConflictingAttrs` look into subqueries?

Reply:
No. Subquery can be done in its own analysis phase?

Reply:
Oh...this change would break the case like:

```scala
Project [a#1]
  Join Inner
    Relation [a#1]
    Relation [a#1]
```
where the plan itself doesn't have conflicting attributes but has a self-join below. Thus, adding `hasConflictingAttrs` would miss self-join resolving. cc @maropu

Reply:
OK let's always dedup relations.